File: plurkconnection/src/main/java/tw/shounenwind/plurkconnection/responses/ApiResponseStream.java
package tw.shounenwind.plurkconnection.responses;

import com.google.common.io.CharStreams;

import java.io.Reader;

public class ApiResponseStream implements IResponse<Reader> {
    public final Reader content;
    public final int statusCode;
    private Throwable e;

    public ApiResponseStream(int statusCode, Reader content) {
        this.statusCode = statusCode;
        this.content = content;
    }

    public ApiResponseStream(Throwable e) {
        this.statusCode = -1;
        this.content = null;
        this.e = e;
    }

    public Throwable getE() {
        return e;
    }

    @Override
    public String toString() {
        try {
            return statusCode + ", " + (content == null ? "" : CharStreams.toString(content));
        } catch (Exception e1) {
            return statusCode + ", " + e1.toString();
        }
    }

    @Override
    public Reader getContent() {
        return content;
    }

    @Override
    public int getStatusCode() {
        return statusCode;
    }
}

File: library/src/main/java/tw/shounenwind/plurkconnection/responses/ApiResponseStream.kt
package tw.shounenwind.plurkconnection.responses

import java.io.Reader

class ApiResponseStream : IResponse<Reader?> {
    override val content: Reader?
    override val statusCode: Int
    val e: Throwable?

    constructor(statusCode: Int, content: Reader) {
        this.statusCode = statusCode
        this.content = content
        this.e = null
    }

    constructor(e: Throwable) {
        this.statusCode = -1
        this.content = null
        this.e = e
    }

    override fun toString(): String {
        return try {
            val stringBuilder = StringBuilder("$statusCode, ")
            if (content != null) {
                stringBuilder.append(content.readText())
            }
            stringBuilder.toString()
        } catch (e1: Exception) {
            "$statusCode, $e1"
        } finally {
            content?.close()
        }

    }
}
